service/src/test/java/bio/terra/pipelines/service/PipelinesServiceTest.java

```diff
@@ -170,4 +189,212 @@ void updatePipelineWorkspaceId() {
+    String notNullErrorMessage = "input_name must not be null";
+    String notEmptyErrorMessage = "input_name must not be empty";
+    String emptyArrayErrorMessage = "input_name must not be an empty array";
+
+    String[] stringArrayWithNull = new String[] {null, "has a null value"};
+
+    return Stream.of(
+        // arguments are: type specification, value to be tested, isRequired, expected errorMessage
+        // basic type checks
+        arguments("Integer", "I am a string", true, commonTypeErrorMessage),
+        arguments("Integer", 2.3, true, commonTypeErrorMessage),
+        arguments("Integer", "2.3", true, commonTypeErrorMessage),
+        arguments("String", "", true, notEmptyErrorMessage),
+        arguments("String", "", false, notEmptyErrorMessage),
+        arguments(
+            "String", List.of("this", "is", "not", "a", "string"), true, commonTypeErrorMessage),
+        arguments("VCF", "path/to/file.vcf", true, vcfTypeErrorMessage),
+        arguments("VCF", 3, true, vcfTypeErrorMessage),
+        arguments(
+            "Array_VCF",
+            new String[] {"path/to/file.vcf.gz", "not a path"},
+            true,
+            arrayVCFTypeErrorMessage),
+        arguments("Array_String", "I am a string", true, commonTypeErrorMessage),
+        arguments(
+            "Array_String",
+            List.of("", "there's an empty string"),
+            true,
+            arrayStringTypeErrorMessage),
+        arguments(
+            "Array_String",
+            new String[] {null, "list with null"},
+            true,
+            arrayStringTypeErrorMessage),
+        arguments("Integer", "I am a string", false, commonTypeErrorMessage),
+        // null and empty checks
+        arguments("String", null, false, notNullErrorMessage),
+        arguments("Integer", null, true, notNullErrorMessage),
+        arguments("Array_String", new String[] {}, true, emptyArrayErrorMessage),
+        arguments("Array_VCF", new String[] {}, true, emptyArrayErrorMessage),
+        arguments("Array_String", null, true, notNullErrorMessage),
+        arguments("Array_String", stringArrayWithNull, true, arrayStringTypeErrorMessage),
+        arguments("Array_VCF", null, true, notNullErrorMessage),
+        arguments(
+            "Array_VCF",
+            new String[] {null, "list/with/null.vcf.gz"},
+            true,
+            arrayVCFTypeErrorMessage));
+  }
+
+  @ParameterizedTest
+  @MethodSource("inputTypeValidationsNotValid")
+  void validateInputTypeNotValid(
+      String type, Object inputValue, Boolean isRequired, String errorMessage) {
+    PipelineInputDefinition inputDefinition =
+        new PipelineInputDefinition(1L, "input_name", type, isRequired);
+    List<PipelineInputDefinition> inputDefinitions = new ArrayList<>(List.of(inputDefinition));
+
+    LinkedHashMap<String, Object> inputs = new LinkedHashMap<>();
+    inputs.put("input_name", inputValue);
+
+    assertEquals(
```

Comment:
another way you could have written this test (or really combining the two above tests) is just asserting whether or not the response from `pipelinesService.validateInputTypes()` was empty or not.  You've already tested the expected error message in the tests above and here you're just making sure what you expect to be valid is valid and what you expect to be invalid is invalid.

so your input arguments would instead be
`String type, Object inputValue, Boolean isRequired, Boolean errorExpected`

Reply:
oh that's nice, I like it. thanks!